keps/sig-node/2535-ensure-secret-pulled-images/README.md

```diff
@@ -165,57 +201,367 @@ ensure all tenants have rights to the images that are already present on a host.
+2. **After the pull**
+    - Once we know the result of the pull operation, we either remove the image pull intent record
+      (=> failed pull attempt), or we write a new "image pulled record" about the image
+      being pulled successfully and remove the original record of the image pull intent.
+    - Without this, if kubelet failed during an image pull, an existing image pull intent record
+      would already grant image access irregardless of this pull operation result for any next
+      pull attempts.
+
+The kubelet uses the following Golang interface to interact with the cache:
+```go
+// ImagePullManager keeps the state of images that were pulled and which are
+// currently still being pulled.
+// It should keep an internal state of images currently being pulled by the kubelet
+// in order to determine whether to destroy a "pulling" record should an image
+// pull fail.
+type ImagePullManager interface {
+    // RecordPullIntent records an intent to pull an image and should be called
+    // before a pull of the image occurs.
+    //
+    // RecordPullIntent() should be called before every image pull. Each call of
+    // RecordPullIntent() must match exactly one call of RecordImagePulled()/RecordImagePullFailed().
+    //
+    // `image` is the content of the pod's container `image` field.
+    RecordPullIntent(image string) error
+    // RecordImagePulled writes a record of an image being successfully pulled
+    // with ImagePullCredentials.
+    //
+    // If `credentials` is nil, it means the image was pulled successfully, but
+    // a reverification should be required upon next pull request. This can be
+    // useful in cases where the kubelet crashed in the middle of the pull but
+    // the image appears on the node after the kubelet restarts.
+    //
+    // `image` is the content of the pod's container `image` field.
+    RecordImagePulled(image, imageRef string, credentials *imagemanager.ImagePullCredentials) error
+    // RecordImagePullFailed should be called if an image failed to pull.
+    //
+    // Internally, it lowers its reference counter for the given image. If the
+    // counter reaches zero, the pull intent record for the image is removed.
+    //
+    // `image` is the content of the pod's container `image` field.
+    RecordImagePullFailed(image string) error
+    // ShouldReverifyCredentials evaluates the policy for image specified in
+    // "image" and if the policy demands reverification, it checks the internal
+    // cache to see if there's a record of pulling the image with the presented
+    // set of credentials or if the image can be accessed by any of the node's pods.
+    //
+    // Returns true if the policy demands reverification and no record of the pull
+    // was found in the cache.
+    //
+    // `image` is the content of the pod's container `image` field.
+    ShouldReverifyCredentials(image, imageRef string, credentials ... *imagemanager.ImagePullSecret) bool
```

Comment:
there's internal verification that happens inside this method where the credentials are compared to the cache, but what this is *really* saying if it returns true is that a pull must be attempted

this might be a clearer name:

```suggestion
    MustAttemptImagePull(image, imageRef string, credentials []imagemanager.ImagePullSecret) bool
```